### PR TITLE
Store keys and values seperately

### LINKED ISSUE #12: Implement `Base.getindex(t::TallyT, k)`

### PR DESCRIPTION
**Pull request by BeastyBlacksmith**
I haven't thought through every implication of this, but I think this makes implementation of dictionary behaviour easier.

I implemented `get`, `+` and `-` with this and you get `getindex` for free then, so fixes #12 .

### CODE REVIEW

File: src/Tally.jl [hunk 1]
```diff
@@ -262,42 +260,70 @@ end
...
 function Base.append!(T::TallyT, it)
   for x in it
     _push!(T, x)
   end
-  sort!(T.data, by = x -> x[2], rev = true)
+  sort!(T)
   return T
 end
 
+function Base.get(T::TallyT, key, default)
```
Comment (thofma) [inline src/Tally.jl:309]: Is this required for the dictionary interface? I would have thought that `getindex` is enough? Or do we get more methods for free when implementing `get`?

**Comment by thofma** [comment]
I pushed some small tweaks.

File: src/Tally.jl (hunk 1 above)
Comment (BeastyBlacksmith) [reply src/Tally.jl:309]: for AbstractDictionaries, `getindex` calls `get`, so we only need that

File: src/Tally.jl [hunk 2]
```diff
@@ -307,21 +307,40 @@ function Base.append!(T::TallyT, it)
...
+  !(T1.by === T2.by && T1.equivalence === T2.equivalence) &&
+    error("Tallies incompatible")
+  K = copy(keys(T1))
+  for k in keys(T2)
+    if !_has_key(k, K, T1.by, T1.equivalence)[1]
+      push!(K, k)
+    end
+  end
+  # To get the right array type
+  V = empty!(vcat(T1.values, T2.values))
```
Comment (BeastyBlacksmith) [inline src/Tally.jl:328]: `V = promote_eltype(T1.values, T2.values)[]`

should be less expensive

**Comment by thofma** [comment]
Thanks!

**Pull request merged**